### PR TITLE
 Make ServiceData thread-safe

### PR DESCRIPTION
**Pull request by Yadunund**
Thread-safe access to the `rmw_service_t` handler. Followed the exact same pattern as in #288

### CODE REVIEW

**Comment by ahcorde** [comment]
[ros2.repos](https://gist.github.com/ahcorde/83e99e600f1c139d7de945f8519c9229)

# rmw_zenoh/yadu/raii-service
* rmw_zenoh commit: https://github.com/ros2/rmw_zenoh/commit/cc4eb19897edfa5c80647b2be782ee2b39724981

| ROS Package      | Failing test name | Failing test output     |
| :---:        |    :----:   |          :---: |
| rcl | None | None |
| rcl_action | None | None |
| rcl_lifecycle   | None        | None      |
| rcl_yaml_param   | None        | None      |
| rclcpp   | test_intra_process_manager        | Failed (Check logs)      |
| rclcpp   | test_node_interfaces__node_graph        | Failed (Check logs)      |
| rclcpp   | test_publisher        | Failed (Check logs)      |
| rclcpp   | test_events_executor        | Timeout (Check logs)      |
| rclcpp   | test_wait_set        | Failed (Check logs)      |
| rclcpp_action   | None       | None      |
| rclcpp_components   | None        | None      |
| rclcpp_lifecycle  | None | -  |
| rmw_zenoh_cpp   | None        | None      |
| test_cli   | None        | None      |
| test_cli_remap   | test_cli_remapping        | TBD      |

File: rmw_zenoh_cpp/src/detail/rmw_data_types.cpp [hunk 1]
```diff
@@ -35,137 +35,6 @@
 ///=============================================================================
 namespace rmw_zenoh_cpp
 {
-///=============================================================================
```
Comment (ahcorde) [inline rmw_zenoh_cpp/src/detail/rmw_data_types.cpp:38]: I think we can remove from the headers

`<optional>`, `<sstream>` and `<string>`

File: rmw_zenoh_cpp/src/detail/rmw_node_data.hpp [hunk 2]
```diff
@@ -76,10 +77,25 @@ class NodeData final
...
+  // Create a new ServiceData for a publisher in this node.
+  bool create_service_data(
+    const rmw_service_t * const service,
+    z_session_t session,
+    std::size_t id,
+    const std::string & service_name,
+    const rosidl_service_type_support_t * type_support,
+    const rmw_qos_profile_t * qos_profile);
+
+  /// Retrieve the ServiceData for a given rmw_subscription_t if present.
```
Comment (ahcorde) [inline rmw_zenoh_cpp/src/detail/rmw_node_data.hpp]:
```suggestion
  // Retrieve the ServiceData for a given rmw_subscription_t if present.
```

File: rmw_zenoh_cpp/src/detail/rmw_node_data.hpp [hunk 3]
```diff
@@ -76,10 +77,25 @@ class NodeData final
     const rmw_qos_profile_t * qos_profile);
 
   /// Retrieve the SubscriptionData for a given rmw_subscription_t if present.
```
Comment (ahcorde) [inline rmw_zenoh_cpp/src/detail/rmw_node_data.hpp]:
```suggestion
  // Retrieve the SubscriptionData for a given rmw_subscription_t if present.
```

**Review by ahcorde — APPROVED** [review]
Just some minor comments. Tests are in the same status, I think it's safe to merge it

**Comment by ahcorde** [comment]
This PR https://github.com/ros2/rclcpp/pull/2648 removed the wait_set failure

**Review by clalancette — APPROVED** [review]
I also tested this with the Nav2 test, and it seems to work pretty well.

**Pull request merged**